Cache dfs path in propose_partitions and re-use that later when trying to find cycles in the graph

tarun292 opened the pull request:
Summary:
This diff introduces a caching mechanism to improve the performance of the partitioner in PyTorch. The changes involve adding a cache to store the DFS path of each node in the graph, which can be reused later when trying to find cycles in the graph.

This shows significant improvements for the edge use cases where the ASR model (which is around 6000+ nodes) used to take 26 minutes, but after this it takes around 8 minutes.

Test Plan: Relying on the existing ExecuTorch CI tests that heavily use this partitioning mechanism and also tested out locally via Bento notebooks.

Differential Revision: D51289200

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -69,38 +111,33 @@ def maybe_merge_partition(self_id: int, other_id: int):
+                        visited_partition_ids.add(partition_id)
 
                 return False
 
```
I think the line 108~113 can be rewritten to only call `dfs_iter_find_cycle` once. 
- gather all user_nodes of "merged_nodes"
- pass this list into `dfs_iter_find_cycle`

```
            for node in merged_nodes:
                for user_node in node.users:
                    if user_node not in merged_nodes:
                        all_user_nodes.appen(user_node)   
                if dfs_iter_find_cycle(all_user_node):
                        # return false indicating cyclic dependency found and
                        # merge is aborted
                        return False
```

tarun292 replied:
That's a good idea. Made the change.

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -1,6 +1,11 @@
+import logging
+from collections import defaultdict, deque
+from copy import copy
+from functools import lru_cache
```
lru_cache is not used anywhere

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -1,6 +1,11 @@
 from torch.fx.passes.utils.fuser_utils import fuse_by_partitions
+import logging
+from collections import defaultdict, deque
+from copy import copy
```
duplicated import

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -1,6 +1,11 @@
 from typing import Dict, List, Set, Iterable, Sequence, Optional, Deque
 
 from torch.fx.passes.utils.fuser_utils import fuse_by_partitions
+import logging
```
duplicated import

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -1,6 +1,11 @@
 
 from torch.fx.passes.utils.fuser_utils import fuse_by_partitions
+import logging
+from collections import defaultdict, deque
```
duplicated import

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -132,6 +172,27 @@ def merge_single_node(node: Node, id: Optional[int]):
+                        partition_map[source_id].add(id)
             else:
                 assignment[node] = id
                 partitions_by_id[id].add_node(node)
```
Do you also need to update partition_map in this branch? 

IIUC, whenever a node assignment is done, you would need to update partition_map ?

tarun292 replied:
Good catch, i need to address the case where partition id is updated for a single node. Let me fix that.

tarun292 replied:
Fixed this in latest version.

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -55,7 +56,44 @@ def __is_node_supported(self, node: Node) -> bool:
+    # for each node. During these traversals it maintains two dictionaries one which keeps
+    # track of all the nodes that are encountered in a DFS traversal starting at a particular
+    # node and another dictionary which keeps track of all the nodes that can reach a particular node.
+    def _cache_dfs_path(self):
```
Effectively, you are maintaining two "dependency_map" for every pair of nodes in the graph. 

I think this is a very useful tool in general, and it worth refactoring this into a util class. 

```
class DependencyViewer:
    def __init__(self, graph_module):
        self.upstreams = defaultdict(set)  # aka dfs_paths_to_node
        self.downstreams = defaultdict(set) # aka dfs_paths_from_node
        
    def depends_on(node_a, node_b) -> bool: 
        ...
    
    def downstreams_of(node):
         ...

    def upstreams_of(node): 
         ...
```

tarun292 replied:
Good idea!

SherlockNoMad replied:
I realized I have implemented this before 
https://www.internalfb.com/code/fbsource/[2f71ada0401e]/fbcode/pytorch/pippy/pippy/fx/passes/infra/partitioner.py?lines=48

Somehow this is not master ??!!

tarun292 replied:
Haha that's very interesting! Very similar approach.

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -55,7 +56,44 @@ def __is_node_supported(self, node: Node) -> bool:
+        dfs_paths_to_node = defaultdict(set)
+        stack = deque()
+
+        for node in self.graph_module.graph.nodes:
```
I think there are better way to build the dependency map. 

You should be able to build the map by just traversing the graph in topological order twice, once in forward order, once in reversal.

SherlockNoMad replied:
upstream[node] = union(node's upstreams' upstream)
downstream[node] = union(node's downstreams' downstream)

tarun292 replied:
Updated. This was a great idea.

albanD commented on docs/source/conf.py:
```diff
@@ -2912,6 +2912,7 @@
     "size_bytes",
     # torch.fx.passes.infra.partitioner
     "CapabilityBasedPartitioner",
+    "DependencyViewer",
```
Adding new entries to this file is not ok.
You need to either properly document this class on the website by adding it to the appropriate .rst or make it a private API.

albanD replied:
Thanks.

SherlockNoMad reviewed (approved):
Looks good to me. Thanks for optimizing this. 
I think it would also be good to add your model as a test case. Existing models in CI are too small to expose the perf problem you ran into.

tarun292 committed:
Cache dfs path in propose_partitions and re-use that later when trying to find cycles in the graph (#115943)

Summary:

This diff introduces a caching mechanism to improve the performance of the partitioner in PyTorch. The changes involve adding a cache to store the DFS path of each node in the graph, which can be reused later when trying to find cycles in the graph.

This shows significant improvements for the edge use cases where the ASR model (which is around 6000+ nodes) used to take 26 minutes, but after this it takes around 8 minutes.

Test Plan: Relying on the existing ExecuTorch CI tests that heavily use this partitioning mechanism and also tested out locally via Bento notebooks.

Reviewed By: SherlockNoMad

Differential Revision: D51289200
